### PR TITLE
Graphql - make schema generation thread safe

### LINKED ISSUE #3934: [GraphQL] Schema generation is not thread safe
Currently multiple threads can try and create the schema until it is cached, this can cause exceptions to be thrown as it causes duplicate keys in some dictionaries.

### PR DESCRIPTION
**Pull request by carlwoodhouse**
fixes #3934

// cc @jrestall

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs [hunk 1]
```diff
@@ -25,64 +27,84 @@ public class SchemaService : ISchemaFactory
...
+        public async Task<ISchema> GetSchema()
         {
-            return _memoryCache.GetOrCreateAsync("GraphQLSchema", async f =>
+            if (_memoryCache.TryGetValue<ISchema>("GraphQLSchema", out var schema)) {
+                return schema;
+            }
+
+            await _schemaGenerationSemaphore.WaitAsync();
+
+            if (_memoryCache.TryGetValue("GraphQLSchema", out schema))
```
[the remark this answers is not in the document]
Comment (carlwoodhouse) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs]: heh, yes it is!

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs [hunk 2]
```diff
@@ -25,64 +27,84 @@ public class SchemaService : ISchemaFactory
...
+                    {
+                        schema.RegisterType(type);
+                    }
 
-                foreach (var type in _serviceProvider.GetServices<IObjectGraphType>())
-                {
-                    schema.RegisterType(type);
-                }
-                
                 // Clean Query, Mutation and Subscription if they have no fields
```
[the remark this answers is not in the document]
Comment (carlwoodhouse) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs]: think it must be mobile

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs [hunk 3]
```diff
@@ -25,64 +27,84 @@ public class SchemaService : ISchemaFactory
             _serviceProvider = serviceProvider;
         }
 
-        public Task<ISchema> GetSchema()
+        public async Task<ISchema> GetSchema()
```
[the remark this answers is not in the document]
Comment (carlwoodhouse) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs]: done!

**Comment by carlwoodhouse** [comment]
@jrestall all resolved

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jrestall) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs]: I checked on desktop and the formatting still looks off to me.

Can you see the 3 lines out of alignment here?
https://github.com/OrchardCMS/OrchardCore/blob/b825d8a234d785f830b17583a5d411e40952fac8/src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs#L79-L82

[the remark this answers is not in the document]
Comment (jrestall) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs]: Oh nevermind, your latest commit seems to have fixed it.

**Comment by jrestall** [comment]
LGTM

**Comment by carlwoodhouse** [comment]
@jrestall build fixed!

**Comment by sebastienros** [comment]
Ideally this service would have been a singleton and then the semaphore could have been a local variable instead of static. But in the end the perf impact is only if many tenants create their schema at the same time. So LGTM

**Comment by sebastienros** [comment]
But the service could definitely be a singleton instead. And the schema could just be a local variable, and we wouldn't need a memory cache. I probably wrote this code, so if anyone understands why it's a scoped service please refresh my memory.

**Comment by carlwoodhouse** [comment]
I think nick wrote it :) the Singleton makes sense. I can investigate it in
a follow up pr if you like.

**Comment by jtkech** [comment]
It is a scoped service maybe because it needs a scoped `_serviceProvider`, but this can be easily fixed by using an httpContextAccessor or just use `ShellScope.Services` (now i think it is better).

But if the issue is on the graphQL side because it uses e.g a non thread safe internal static dictionary, then we would still need a global lock.

**Comment by carlwoodhouse** [comment]
@sebastienros @jtkech @jrestall 

As suggested i've converted to a singleton and i'm no longer able to replicate the thread safety issue, uses `ShellScope.Services` as per @jtkech 's suggestion

**Comment by jtkech** [comment]
Good to know but if it is a graphQL issue that may use internally a shared static dico, now that `SchemaService` is a tenant singleton, to be sure we don't need a global lock, we would have to test it with at least 2 different tenants creating their schema concurrently (maybe this is what you already did).

**Comment by carlwoodhouse** [comment]
@jtkech well i just tried to test that and have another issue, not sure why but SchemaService seems to be a global singleton, rather then tenant ? (so it returns the wrong schema for one tenant) i assumed all singletons where tenant level ?

**Comment by carlwoodhouse** [comment]
Ignore my previous comment if you saw it, i've tested @jtkech and its working fine for multiple tenants doing concurrent schema generation

**Comment by jtkech** [comment]
Okay, this is our builders that are not thread safe at the tenant level when concurrent threads use the same `IInputObjectGraphTypes` registered as tenant singletons and e.g act on the same `FieldTypes`.

I could repro by running 2 `GetSchema()` at the same time, and i found a place that is not thread safe (maybe there are others), in `FieldTypeExtensions` when trying to add an item that already exists.

        internal static FieldType WithMetaData(this FieldType fieldType, string name, object value)
        {
            if (fieldType == null)
            {
                return null;
            }

            if (fieldType.Metadata == null)
            {
                fieldType.Metadata = new Dictionary<string, object>();
            }

            if (!fieldType.HasMetadata(name))
            {
                // here it failed to add an item that already exists
                fieldType.Metadata.Add(name, value);
            }

            return fieldType;
        }

So, because the issue is at the tenant level, not from graphQL using e.g a shared static dico, it seems that we don't need a global lock, we just need a tenant level lock as you did.

So LGTM.

**Comment by carlwoodhouse** [comment]
Yeah that was original issue that I came across too that prompted me to
look at it :)

**Comment by Jetski5822** [comment]
ShellScope.Services; Sexy, Approved from me

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs [hunk 4]
```diff
@@ -1,66 +1,76 @@
...
-                    DependencyResolver = _serviceProvider.GetService<IDependencyResolver>()
+                    DependencyResolver = serviceProvider.GetService<IDependencyResolver>()
                 };
 
                 foreach (var builder in _schemaBuilders)
                 {
                     var token = await builder.BuildAsync(schema);
 
-                    if (token != null)
+                    if (token is object)
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs]: `token != null` and everywhere else it's used

Comment (carlwoodhouse) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs]: @sebastienros do you want another pr with that change .. someone merged this one it seems ;p

Comment (Jetski5822) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs]: whats wrong with `is object` SonarQube allow it :P

**Pull request merged**